decentralized-message-queue/src/DMQ/NodeToNode.hs

```diff
@@ -0,0 +1,511 @@
+import Control.Monad.Class.MonadST (MonadST)
+import Control.Monad.Class.MonadThrow (MonadMask, MonadThrow)
+import Control.Monad.Class.MonadTimer.SI (MonadTimer)
+import Control.Tracer (Tracer, nullTracer)
+import Data.ByteString.Lazy qualified as BL
+import Data.Hashable (Hashable)
+import Data.Void (Void)
+import DMQ.Diffusion.NodeKernel (NodeKernel (..))
+import Network.Mux.Types (Mode (..))
+import Network.TypedProtocol.Codec (Codec)
+import Ouroboros.Network.BlockFetch.ClientRegistry (bracketKeepAliveClient)
+import Ouroboros.Network.Channel (Channel)
+import Ouroboros.Network.CodecCBORTerm (CodecCBORTerm (..))
+import Ouroboros.Network.ConnectionId (ConnectionId (..))
+import Ouroboros.Network.Driver.Limits (runPeerWithLimits)
+import Ouroboros.Network.Handshake.Acceptable (Acceptable (..))
+import Ouroboros.Network.Handshake.Queryable (Queryable (..))
+import Ouroboros.Network.KeepAlive (KeepAliveInterval (..), keepAliveClient,
+           keepAliveServer)
+import Ouroboros.Network.Magic (NetworkMagic (..))
+import Ouroboros.Network.Mux (MiniProtocol (..), MiniProtocolCb (..),
+           MiniProtocolLimits (..), OuroborosBundle,
+           OuroborosBundleWithExpandedCtx, RunMiniProtocol (..),
+           StartOnDemandOrEagerly (..), TemperatureBundle (..),
+           WithProtocolTemperature (..))
+import Ouroboros.Network.NodeToNode (DiffusionMode (..),
+           ExpandedInitiatorContext (..), HandshakeTr,
+           NodeToNodeVersionData (..), ResponderContext (..), addSafetyMargin,
+           keepAliveMiniProtocolNum, peerSharingMiniProtocolNum)
+import Ouroboros.Network.NodeToNode.Version qualified as NTN
+import Ouroboros.Network.PeerSelection (PeerSharing (..))
+import Ouroboros.Network.PeerSharing (bracketPeerSharingClient,
+           peerSharingClient, peerSharingServer)
+import Ouroboros.Network.Protocol.Handshake (HandshakeArguments (..))
+import Ouroboros.Network.Protocol.Handshake.Codec (cborTermVersionDataCodec,
+           codecHandshake, timeLimitsHandshake)
+import Ouroboros.Network.Protocol.KeepAlive.Client (keepAliveClientPeer)
+import Ouroboros.Network.Protocol.KeepAlive.Codec (byteLimitsKeepAlive,
+           codecKeepAlive_v2, timeLimitsKeepAlive)
+import Ouroboros.Network.Protocol.KeepAlive.Server (keepAliveServerPeer)
+import Ouroboros.Network.Protocol.KeepAlive.Type (KeepAlive)
+import Ouroboros.Network.Protocol.Limits (ProtocolSizeLimits,
+           ProtocolTimeLimits)
+import Ouroboros.Network.Protocol.PeerSharing.Client (peerSharingClientPeer)
+import Ouroboros.Network.Protocol.PeerSharing.Codec (byteLimitsPeerSharing,
+           codecPeerSharing, timeLimitsPeerSharing)
+import Ouroboros.Network.Protocol.PeerSharing.Server (peerSharingServerPeer)
+import Ouroboros.Network.Protocol.PeerSharing.Type qualified as Protocol
+import System.Random (mkStdGen)
+
+data NodeToNodeVersion =
+  NodeToNodeV_1
+  deriving (Eq, Ord, Enum, Bounded, Show, Generic, NFData)
+
```

Comment:
Double check in this module if we are going to need `NodeToNodeVersionData` `diffusionMode` parameter, since DMQ Node will only be run be SPOs

Reply:
We should create our own `NodeToNodeVersionData` here to not be tied to ouroboros-network one